app/models/voucher.rb

```diff
@@ -1,3 +1,66 @@
 class Voucher < ApplicationRecord
+  VOUCHER_ATTRIBUTES = %i(
+    expiry_date
+    code
+    percent_discount
+    min_total_price
+  ).freeze
+
   has_many :booking_vouchers, dependent: :destroy
+  validate :validate_expiry_date
+  validate :validate_unique_code, on: %i(new update)
+
+  scope :ordered, (lambda do
+    select("vouchers.*,
+            CASE WHEN expiry_date >= NOW() THEN 0 ELSE 1 END AS order_field")
+    .order("order_field, created_at ASC")
+  end)
+  scope :available, (lambda do
+    where("expiry_date >= NOW()").order(expiry_date: :asc)
+  end)
+  scope :expired, (lambda do
+    where("expiry_date < NOW()").order(expiry_date: :asc)
+  end)
+  scope :by_status, lambda{|status|
+    case status
+    when "all"
+      ordered
+    when "available"
+      available
+    when "expired"
+      expired
+    else
+      all
+    end
+  }
+  scope :by_code, lambda {|code|
+    where("code LIKE ?", "%#{sanitize_sql_like(code)}%") if code.present?
+  }
```

Comment:
```suggestion
  scope :by_code, lambda do |code|
    where("code LIKE ?", "%#{sanitize_sql_like(code)}%") if code.present?
  end
```
Dùng do end cho nhiều dòng em nhé